File: tasks/bench-lease-lock/fixtures/reference/lease.py
import sys
from collections import deque

class Res:
    def __init__(self):
        self.holder = None
        self.deadline = None
        self.q = deque()

def release(res, tick):
    res.holder = None
    res.deadline = None
    if res.q:
        who, ln = res.q.popleft()
        res.holder, res.deadline = who, tick + ln

def main(path):
    rs = {f"R{i}": Res() for i in range(1, 6)}
    for raw in open(path):
        raw = raw.strip()
        if not raw or raw.startswith('#'):
            continue
        f = raw.split()
        tick, c, act, name = int(f[0]), f[1], f[2], f[3]
        n = int(f[4]) if len(f) > 4 else None
        r = rs[name]
        if r.holder is not None and r.deadline <= tick:
            release(r, tick)
        if act == 'LOCK':
            if r.holder is None:
                r.holder, r.deadline = c, tick + n
            elif r.holder != c and all(w != c for w, _ in r.q):
                r.q.append((c, n))
        elif act == 'UNLOCK':
            if r.holder == c:
                release(r, tick)
        elif act == 'RENEW':
            if r.holder == c:
                r.deadline = tick + n
    parts = []
    for i in range(1, 6):
        r = rs[f"R{i}"]
        st = 'free' if r.holder is None else f"{r.holder}@{r.deadline}"
        parts.append(f"R{i}:{st}[{','.join(w for w, _ in r.q)}]")
    print(' '.join(parts))

main(sys.argv[1])

File: tasks/bench-lease-lock/fixtures/reference/lease.mjs
// Reference simulator for the LEASE spec. `variant` selects a deliberately wrong reading (skip-holder, dedupe, eager), used only to prove the log discriminates.
export function parseLog(text) {
  return text.split('\n').map(l => l.trim()).filter(l => l && !l.startsWith('#')).map(l => {
    const p = l.split(/\s+/);
    return { tick: Number(p[0]), client: p[1], action: p[2], res: p[3], lease: p[4] === undefined ? undefined : Number(p[4]) };
  });
}

export function simulate(events, variant = 'spec') {
  const R = {};
  for (let i = 1; i <= 5; i++) R['R' + i] = { holder: null, deadline: 0, queue: [] };
  const handOver = (r, tick) => {
    r.holder = null;
    if (r.queue.length) { const e = r.queue.shift(); r.holder = e.client; r.deadline = tick + e.lease; }
  };
  const expire = (r, tick) => { if (r.holder !== null && r.deadline <= tick) handOver(r, tick); };
  for (const ev of events) {
    if (variant === 'eager') for (const k of Object.keys(R)) expire(R[k], ev.tick);
    const r = R[ev.res];
    if (variant !== 'eager') expire(r, ev.tick);
    if (ev.action === 'LOCK') {
      if (r.holder === null) { r.holder = ev.client; r.deadline = ev.tick + ev.lease; }
      else if (variant === 'skip-holder' && r.holder === ev.client) { /* wrong: holder cannot queue behind itself */ }
      else if (variant === 'dedupe' && (r.holder === ev.client || r.queue.some(e => e.client === ev.client))) { /* wrong: one request per client */ }
      else r.queue.push({ client: ev.client, lease: ev.lease });
    } else if (ev.action === 'UNLOCK') {
      if (r.holder === ev.client) handOver(r, ev.tick);
    } else if (ev.action === 'RENEW') {
      if (r.holder === ev.client) r.deadline = ev.tick + ev.lease;
    }
  }
  return Object.keys(R).map(k => { const r = R[k]; return `${k}:${r.holder === null ? 'free' : r.holder + '@' + r.deadline}[${r.queue.map(e => e.client).join(',')}]`; }).join(' ');
}
